### ISSUE TITLE
change batch size in mnist example got wrong answer

### ISSUE THREAD

**Issue by fastalgo** [tiny-dnn/tiny-dnn#538]
I changed the batch size of mnist.cpp example from 10 to 64.

https://github.com/tiny-dnn/tiny-dnn/blob/master/examples/mnist/train.cpp

Then I got a wrong answer. Why?

**Comment by pliptor** [comment]
@Randl Could you point me out where to read on the relation between dropout and batch sizes? Thanks!

**Comment by Randl** [comment]
@pliptor As for this bug, I suppose the proble is in batch and not in dropout, just with droupout it causes crash and with different settings it causes wrong results.

Generally I believe dropout and batch size are not very related.

**Comment by pliptor** [comment]
Thank you. It seems both issues are very important to be solved and I wanted to learn more about them. Let me know if there's anything I could help with.

**Comment by Randl** [comment]
We need to understand what causes the bug and to localize it first. If you have any insights, feel free to share them.

**Comment by pliptor** [comment]
I tried to look at the dropout code earlier but no clue so far. Will keep trying.

**Comment by bhack** [comment]
@Randl I think it is safer to have a test on the batch size dimension

**Comment by pliptor** [comment]
The learning rate of the optimizer gets multiplied by sqrt(batch dimension) in the mnist code. For a batch size of 64 that's 8. It might be too fast. Let me see.

**Comment by fastalgo** [comment]
@pliptor Thanks so much!

**Comment by pliptor** [comment]
@fastalgo  My machine is slow so I can't simulate fast I'm getting good results with batch size of 64 by limiting the multiplying factor with std::max(float_t(4),...). The 4 came from the fact that a size of 16 was working fine for me. But this is a hack. There might be better things to do here.

**Comment by fastalgo** [comment]
@pliptor I got it, have a nice day :-)

**Comment by Randl** [comment]
@pliptor Well those issues might be not connected after all.

As for limiting learning rate, I don't see something much better than that.

**Comment by pliptor** [comment]
It's a hack but I can submit a patch for mnist if you want.

**Comment by Randl** [comment]
If you can open PR, we'll discuss it there.

**Comment by Randl** [comment]
@pliptor Probably, just add it to https://github.com/tiny-dnn/tiny-dnn/pull/574

**Comment by pliptor** [comment]
Good idea.

**Comment by Randl** [comment]
@fastalgo Can you check with latest master?

**Comment by fastalgo** [comment]
@Randl Yes, it is working, thanks!
